catalog/app/containers/Bucket/ListingActions.tsx

```diff
@@ -0,0 +1,240 @@
+
+import * as Bookmarks from 'containers/Bookmarks/Provider'
+import * as Model from 'model'
+import * as AWS from 'utils/AWS'
+import * as NamedRoutes from 'utils/NamedRoutes'
+import * as s3paths from 'utils/s3paths'
+
+import * as FileView from './FileView'
+
+const useButtonStyles = M.makeStyles({
+  root: {
+    padding: '5px',
+  },
+})
+
+interface ButtonProps extends M.IconButtonProps {
+  download?: boolean
+  href?: string
+  icon: string
+}
+
+function Button({ icon, className, ...props }: ButtonProps) {
+  const classes = useButtonStyles()
+  return (
+    <M.IconButton className={cx(classes.root, className)} {...props}>
+      <M.Icon>{icon}</M.Icon>
+    </M.IconButton>
+  )
+}
+
+interface BucketButtonProps {
+  className: string
+  location: Model.S3.S3ObjectLocation
+}
+
+function Bookmark({ className, location }: BucketButtonProps) {
+  const bookmarks = Bookmarks.use()
+  const isBookmarked = location ? bookmarks.isBookmarked('main', location) : false
+  const toggleBookmark = () => location && bookmarks.toggle('main', location)
+  return (
+    <Button
+      icon={isBookmarked ? 'turned_in' : 'turned_in_not'}
+      className={className}
+      title="Bookmark"
+      onClick={toggleBookmark}
+    />
+  )
+}
+
+function BucketDirectory({ className, location: { bucket, key } }: BucketButtonProps) {
+  return (
+    <FileView.ZipDownloadForm suffix={`dir/${bucket}/${key}`} className={className}>
+      <Button icon="arrow_downward" title="Download" type="submit" />
+    </FileView.ZipDownloadForm>
+  )
+}
+
+function BucketFile({ className, location }: BucketButtonProps) {
+  return AWS.Signer.withDownloadUrl(location, (url: string) => (
+    <Button
+      href={url}
+      className={className}
+      title="Download"
+      download
+      icon="arrow_downward"
+    />
+  ))
+}
+
+interface PackageDownloadButtonProps {
+  className: string
+  handle: {
+    bucket: string
+    name: string
+  }
+  revision: string
+  path?: string
+}
+
+function PackageDownloadButton({
+  className,
+  handle: { bucket, name },
+  revision,
+  path,
+}: PackageDownloadButtonProps) {
+  const downloadPath = path
+    ? `package/${bucket}/${name}/${revision}/${path}`
+    : `package/${bucket}/${name}/${revision}`
+  return (
+    <FileView.ZipDownloadForm suffix={downloadPath} className={className}>
+      <Button icon="arrow_downward" title="Download" type="submit" />
+    </FileView.ZipDownloadForm>
+  )
+}
+
+const useRowActionsStyles = M.makeStyles((t) => ({
+  // TODO: add palette.paper overlay and then use palette color for container
+  // bgReset: {},
+  root: {
+    height: '100%',
+    display: 'flex',
+    flexDirection: 'row-reverse',
+  },
+  container: {
+    '.MuiDataGrid-row:hover &': {
+      animation: `$show 150ms ease-out`,
+      position: 'absolute',
+    },
+    display: 'flex',
+    background: `linear-gradient(to right, ${fade('#f5f5f5', 0)} 0, #f5f5f5 ${t.spacing(
+      4,
```

Comment:
oh how i hate what prettier does in such cases :rage4:

Reply:
I formatted it to multiline notation

Reply:
oh, that actually looks better